core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java

```diff
@@ -3413,6 +3413,54 @@ RelOptFixture checkDynamicFunctions(boolean treatDynamicCallsAsConstant) {
         .check();
   }
 
+  @Test void testEmptyTableProject() {
+    // table is transformed to empty values and extra project will be removed.
+    final String sql = "select * from EMPTY_PRODUCTS\n";
+    sql(sql)
+        .withRule(
+            PruneEmptyRules.EMPTY_TABLE,
+            PruneEmptyRules.PROJECT_INSTANCE)
+        .check();
+  }
+
+  @Test void testEmptyTableJoinLeft() {
+    // inner join is eliminated after the left table is transformed to empty values.
+    final String sql = "select * from EMPTY_PRODUCTS as e\n"
+        + ",products as d where e.PRODUCTID = d.PRODUCTID\n";
+    sql(sql)
+        .withRule(
+            PruneEmptyRules.EMPTY_TABLE,
+            PruneEmptyRules.JOIN_LEFT_INSTANCE)
+        .check();
+  }
+
+  @Test void testEmptyTableJoinRight() {
+    // inner join is eliminated after the right table is transformed to empty values.
+    final String sql = "select * from products as e\n"
+        + ",EMPTY_PRODUCTS as d where e.PRODUCTID = d.PRODUCTID\n";
+    sql(sql)
+        .withRule(
+            PruneEmptyRules.EMPTY_TABLE,
+            PruneEmptyRules.JOIN_RIGHT_INSTANCE)
+        .check();
+  }
```

Comment:
I don't think need a test case for every possible combination of rules of `EMPTY_TABLE` with the others. I think we can keep just two:
* `testEmptyTable` which just tests the new rule.
* `testEmptyTableInComplexQuery` which exploits multiple pruning rules including `EMPTY_TABLE` to simplify a plan completely leaving only an empty `LogicalValues` at the end.